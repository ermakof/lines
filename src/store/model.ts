import IUserProfile from '@src/modules/Auth/model/IUserProfile';

export interface IState {
  isLoading?: boolean;
  gameFieldSize: number;
  gameFieldPercentFilled: number;
  gameFieldData: Array<number>;
  selectedCell?: number;
  userLevel: string;
  userProfile?: IUserProfile;
}
