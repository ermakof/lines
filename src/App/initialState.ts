import { IAppState } from '@src/App/model/IAppState';
import createGameField from '@src/utils/createGameField';

export const GAME_FIELD_SIZE = 9;

export const INIT_GAME_LEVEL = '1';
export const INIT_GAME_FIELD_PERCENT_FILLED = 0;

interface ISettings {
  percent: number;
  chain: number;
  increase: number;
}
interface IProps {
  [key: string]: ISettings;
}
export const GAME_LEVEL_SETTINGS: IProps = {
  '1': { percent: 10, chain: 2, increase: 3 },
  '2': { percent: 20, chain: 3, increase: 4 },
  '3': { percent: 30, chain: 4, increase: 5 },
};

const initialState: IAppState = {
  userLevel: INIT_GAME_LEVEL,
  gameFieldPercentFilled: 0,
  gameFieldData: createGameField(GAME_FIELD_SIZE ** 2, INIT_GAME_FIELD_PERCENT_FILLED),
  highlightedCells: undefined,
  score: 0,
  selectedCell: undefined,
  isLoading: false,
  hitParade: undefined,
};

export default initialState;
