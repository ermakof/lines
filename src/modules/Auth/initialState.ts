import { IAuthState } from '@src/modules/Auth/model/IAuthState';

const initialState: IAuthState = {
  userProfile: undefined,
};

export default initialState;
