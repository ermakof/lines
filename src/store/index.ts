import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';

import rootReducer from '@src/store/rootReducer';

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware()
      .prepend
      // correctly typed middlewares can just be used
      ()
      // prepend and concat calls can be chained
      .concat(logger),
});

type TRootState = ReturnType<typeof store.getState>;

export { store, TRootState };
