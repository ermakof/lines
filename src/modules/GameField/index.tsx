import React, { FC } from 'react';
import styled from '@emotion/styled';

import { ICellInfo } from '@src/modules/Cell/models';
import Cell from '@src/modules/Cell';
import { TRootState } from '@src/store';
import { useSelector } from 'react-redux';
import { GAME_FIELD_SIZE } from '@src/App/initialState';
import Section from '@src/layout/Section';

interface IFieldContainer {
  active?: boolean;
}
export const FieldContainer = styled.section<IFieldContainer>`
  margin: auto;
  min-width: 525px;
  z-index: 999;
`;

const Root = styled(Section)`
  background: '#282c34';
`;

const MessageNoData = styled.p`
  height: 80vh;
  position: relative;
  text-align: center;
`;

const GameField: FC = () => {
  const {
    gameFieldData,
    selectedCell,
    highlightedCells = {},
  } = useSelector(({ app }: TRootState) => app);
  const widthMinus1 = GAME_FIELD_SIZE - 1;

  if (!gameFieldData.length) {
    return <MessageNoData role="noDataMessage">Нет данных!</MessageNoData>;
  }

  return (
    <Root>
      <FieldContainer role="cellsGrid">
        {gameFieldData.map((item: ICellInfo, index: number) => {
          const y = Math.floor(index / GAME_FIELD_SIZE);
          const x = index % GAME_FIELD_SIZE;
          const isRight = x === widthMinus1;
          const isLeft = x === 0;
          return (
            <Cell
              key={`${y}-${x}`}
              num={index}
              filled={item}
              isSelected={selectedCell === index}
              isLeft={isLeft}
              isRight={isRight}
              highlighted={highlightedCells[index]}
            />
          );
        })}
        <div style={{ clear: 'both' }} />
      </FieldContainer>
    </Root>
  );
};

export default GameField;
