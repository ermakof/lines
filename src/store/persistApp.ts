import { LOCAL_STORAGE_APP_KEY, store } from '@src/store/index';

const persistApp = () => {
  const app = store.getState().app;
  localStorage.setItem(LOCAL_STORAGE_APP_KEY, JSON.stringify(app));
};

export default persistApp;
