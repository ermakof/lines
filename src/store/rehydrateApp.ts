import { actions } from '@src/modules/Auth/authSlice';
import { store } from '@src/store/index';

export const rehydrateApp = () => {
  store.dispatch(actions.rehydrate());
};
