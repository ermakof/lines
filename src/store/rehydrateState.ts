import appSlice from '@src/App/appSlice';
import { LOCAL_STORAGE_APP_KEY, store } from '@src/store/index';

const rehydrateState = () => {
  try {
    const persistedState = localStorage.getItem(LOCAL_STORAGE_APP_KEY);
    if (persistedState) {
      const payload = JSON.parse(persistedState);
      store.dispatch(appSlice.actions.hydrate(payload));
    }
  } catch (e) {
    console.log(e);
  }
};

export default rehydrateState;
