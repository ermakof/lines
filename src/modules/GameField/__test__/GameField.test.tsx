import React from 'react';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import { useSelector, useDispatch } from 'react-redux';

import GameField from '../index';
import initialState from '@src/App/initialState';
import { TRootState } from '@src/store';

const mockStore = configureMockStore();
let store: TRootState;

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('GameField', () => {
  beforeEach(() => {
    store = mockStore(initialState);
  });

  it('No render cells, game field size = 0', () => {
    const mockDispatch = jest.fn();
    (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
    (useSelector as jest.Mock).mockReturnValue({
      gameFieldData: [],
    });
    const { asFragment } = render(
      <Provider store={store}>
        <GameField />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const message = screen.getByRole(/^noDataMessage$/gi);
    expect(message).toBeInTheDocument();
  });

  it('renders 9 cells from data:[9 items]', () => {
    const mockDispatch = jest.fn();
    (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
    (useSelector as jest.Mock).mockReturnValue({
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
    });
    const { asFragment } = render(
      <Provider store={store}>
        <GameField />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const cellsGrid = screen.getByRole(/cellsGrid/gi);
    expect(cellsGrid).toBeInTheDocument();
    const cellContainer = screen.getAllByRole(/cellContainer/gi);
    expect(cellContainer.length).toBe(5);
  });
});
