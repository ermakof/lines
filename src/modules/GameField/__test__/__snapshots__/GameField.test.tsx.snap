// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameField No render cells, game field size = 0 1`] = `
<DocumentFragment>
  <p
    class="css-5evfw1"
    role="noDataMessage"
  >
    Нет данных!
  </p>
</DocumentFragment>
`;

exports[`GameField renders 9 cells from data:[9 items] 1`] = `
<DocumentFragment>
  <div
    class="css-1kwt2ug"
  >
    <section
      class="css-1v4wim7"
      role="cellsGrid"
    >
      <div
        class="css-1bqu5zx"
        role="cellContainer-0"
      >
        <div
          class="css-zzdbx3"
          role="cellContent-0"
        />
      </div>
      <div
        class="css-g263yg"
        role="cellContainer-1"
      />
      <div
        class="css-g263yg"
        role="cellContainer-2"
      />
      <div
        class="css-g263yg"
        role="cellContainer-3"
      />
      <div
        class="css-g263yg"
        role="cellContainer-4"
      />
      <div
        class="css-g263yg"
        role="cellContainer-5"
      />
      <div
        class="css-g263yg"
        role="cellContainer-6"
      />
      <div
        class="css-g263yg"
        role="cellContainer-7"
      />
      <div
        class="css-v213q6"
        role="cellContainer-8"
      />
      <div
        style="clear: both;"
      />
    </section>
  </div>
</DocumentFragment>
`;
